File: lib/mappers/VariableMapper.ts
/**
 * VariableMapper - 変数データのマッピング
 */

import { BaseMapper } from './BaseMapper';
import { Variable, CreateVariableInput, UpdateVariableInput } from '../types/variable';
import { Logger } from '../logger';

interface VariableRaw {
  id: string;
  name: string;
  type: string;
  label: string | null;
  icon: string | null;
  valid: number;
  createdAt: string;
  updatedAt: string;
}

export class VariableMapper extends BaseMapper<Variable, VariableRaw> {
  /**
   * DB生データをEntityに変換
   */
  toEntity(raw: VariableRaw): Variable {
    return {
      id: raw.id,
      name: raw.name,
      type: raw.type as any,
      label: raw.label || undefined,
      icon: raw.icon || undefined,
      valid: raw.valid === 1,
      createdAt: raw.createdAt,
      updatedAt: raw.updatedAt,
    };
  }

  /**
   * EntityをDB用のデータに変換
   */
  toRaw(entity: Partial<Variable>): VariableRaw {
    return {
      id: entity.id || this.generateId(),
      name: entity.name || '',
      type: entity.type || 'custom',
      label: entity.label || null,
      icon: entity.icon || null,
      valid: entity.valid !== undefined ? (entity.valid ? 1 : 0) : 1,
      createdAt: entity.createdAt || new Date().toISOString(),
      updatedAt: new Date().toISOString(),
    };
  }

  /**
   * 全変数を取得（有効なもののみ）
   */
  getAll(): Variable[] {
    return this.fetchAll('SELECT * FROM variables WHERE valid = 1 ORDER BY name ASC');
  }

  /**
   * 全変数を取得（無効なものも含む）
   * 設定画面での表示用
   */
  getAllIncludingInvalid(): Variable[] {
    return this.fetchAll('SELECT * FROM variables ORDER BY createdAt ASC');
  }

  /**
   * IDで変数を取得
   */
  getById(id: string): Variable | null {
    return this.fetchOne('SELECT * FROM variables WHERE id = ?', [id]);
  }

  /**
   * 名前で変数を取得
   */
  getByName(name: string): Variable | null {
    return this.fetchOne('SELECT * FROM variables WHERE name = ?', [name]);
  }

  /**
   * タイプで変数を取得（有効なもののみ）
   */
  getByType(type: string): Variable[] {
    return this.fetchAll('SELECT * FROM variables WHERE type = ? AND valid = 1 ORDER BY name ASC', [type]);
  }

  /**
   * 変数を作成
   */
  create(input: CreateVariableInput): Variable {
    const raw = this.toRaw({
      name: input.name,
      type: input.type,
      label: input.label,
      icon: input.icon,
    });

    this.insert(
      `INSERT INTO variables (id, name, type, label, icon, valid, createdAt, updatedAt)
       VALUES (?, ?, ?, ?, ?, ?, ?, ?)`,
      [raw.id, raw.name, raw.type, raw.label, raw.icon, raw.valid, raw.createdAt, raw.updatedAt]
    );

    return this.toEntity(raw);
  }

  /**
   * 変数を更新
   */
  updateById(id: string, input: UpdateVariableInput): Variable | null {
    const existing = this.getById(id);
    if (!existing) return null;

    const updatedAt = new Date().toISOString();

    this.update(
      `UPDATE variables
       SET name = ?, label = ?, icon = ?, updatedAt = ?
       WHERE id = ?`,
      [
        input.name ?? existing.name,
        input.label ?? existing.label,
        input.icon ?? existing.icon,
        updatedAt,
        id
      ]
    );

    return this.getById(id);
  }

  /**
   * 変数を削除
   */
  deleteById(id: string): boolean {
    const existing = this.getById(id);
    if (!existing) return false;

    this.delete('DELETE FROM variables WHERE id = ?', [id]);
    return true;
  }

  /**
   * プランに応じてvalidフラグを更新（customタイプのみ、FREE_VARIABLES_LIMIT分だけ有効にする）
   */
  updateValidFlags(limit: number): void {
    try {
      // customタイプの変数をすべて無効にする
      this.update(`UPDATE variables SET valid = 0 WHERE type = 'custom'`, []);

      // createdAt ASCで上位N件を有効にする
      this.update(`
        UPDATE variables
        SET valid = 1
        WHERE id IN (
          SELECT id FROM variables
          WHERE type = 'custom'
          ORDER BY createdAt ASC
          LIMIT ?
        )
      `, [limit]);

      Logger.info(`[VariableMapper] Updated valid flags: ${limit} variables are now valid`);
    } catch (error) {
      Logger.error('[VariableMapper] Failed to update valid flags:', error);
      throw error;
    }
  }
}

export const variableMapper = new VariableMapper();

File: ios/ClipTapKeyboard/Mappers/VariableMapper.swift
//
//  VariableMapper.swift
//  ClipTapKeyboard
//
//  変数Mapper（TypeScript版 VariableMapper.ts と同等）
//  読み取り専用操作のみ実装
//

import Foundation
import SQLite3

class VariableMapper: BaseMapper {

    // MARK: - Singleton

    static let shared = VariableMapper()

    // MARK: - Initialization

    private init() {
        super.init(tableName: "variables")
    }

    // MARK: - Read Operations

    /// 全変数を取得（valid=1のみ）
    func getAll() -> [Variable] {
        let query = """
            SELECT id, name, type, label, icon, valid, sortOrder, createdAt, updatedAt
            FROM \(tableName)
            WHERE valid = 1
            ORDER BY sortOrder ASC
        """

        return executeQuery(query) { statement in
            return self.mapVariable(from: statement)
        }
    }

    /// 全変数を取得（無効なものも含む）
    func getAllIncludingInvalid() -> [Variable] {
        let query = """
            SELECT id, name, type, label, icon, valid, sortOrder, createdAt, updatedAt
            FROM \(tableName)
            ORDER BY sortOrder ASC
        """

        return executeQuery(query) { statement in
            return self.mapVariable(from: statement)
        }
    }

    /// ID指定で変数を取得
    func getById(_ id: String) -> Variable? {
        let query = """
            SELECT id, name, type, label, icon, valid, sortOrder, createdAt, updatedAt
            FROM \(tableName)
            WHERE id = ?
        """

        return executeQuerySingle(query, parameters: [id]) { statement in
            return self.mapVariable(from: statement)
        }
    }

    /// 名前で変数を取得
    func getByName(_ name: String) -> Variable? {
        let query = """
            SELECT id, name, type, label, icon, valid, sortOrder, createdAt, updatedAt
            FROM \(tableName)
            WHERE name = ?
        """

        return executeQuerySingle(query, parameters: [name]) { statement in
            return self.mapVariable(from: statement)
        }
    }

    /// タイプ別に変数を取得
    func getByType(_ type: String) -> [Variable] {
        let query = """
            SELECT id, name, type, label, icon, valid, sortOrder, createdAt, updatedAt
            FROM \(tableName)
            WHERE type = ? AND valid = 1
            ORDER BY sortOrder ASC
        """

        return executeQuery(query, parameters: [type]) { statement in
            return self.mapVariable(from: statement)
        }
    }

    // MARK: - Mapping

    /// SQLite結果からVariableモデルにマッピング
    private func mapVariable(from statement: OpaquePointer) -> Variable {
        let id = getString(statement, at: 0) ?? ""
        let name = getString(statement, at: 1) ?? ""
        let type = getString(statement, at: 2) ?? ""
        let label = getString(statement, at: 3)
        let icon = getString(statement, at: 4)
        let valid = getBool(statement, at: 5)
        let sortOrder = getInt(statement, at: 6)
        let createdAt = getString(statement, at: 7) ?? ""
        let updatedAt = getString(statement, at: 8) ?? ""

        return Variable(
            id: id,
            name: name,
            type: type,
            label: label,
            icon: icon,
            valid: valid,
            sortOrder: sortOrder,
            createdAt: createdAt,
            updatedAt: updatedAt
        )
    }

    // MARK: - Valid Flags Update

    /// validフラグを更新（サブスクリプション状態に応じて）
    ///
    /// 処理内容：
    /// 1. customタイプの変数をすべて無効（valid=0）にする
    /// 2. 作成日時の古い順にlimit件を有効化（valid=1）
    ///
    /// 注意：systemタイプの変数は常に有効なので、更新対象外
    ///
    /// - Parameter limit: 有効にする最大カスタム変数数
    func updateValidFlags(limit: Int) {
        print("[VariableMapper] 🔄 updateValidFlags called with limit: \(limit)")

        guard let db = Database.shared.getDB() else {
            print("[VariableMapper] ❌ Database not available")
            return
        }

        do {
            // 1. customタイプの変数をすべて無効にする
            let updateAllQuery = "UPDATE variables SET valid = 0 WHERE type = 'custom'"
            var statement: OpaquePointer?

            if sqlite3_prepare_v2(db, updateAllQuery, -1, &statement, nil) == SQLITE_OK {
                if sqlite3_step(statement) == SQLITE_DONE {
                    print("[VariableMapper] ✅ Set all custom variables to invalid")
                } else {
                    let errorMsg = String(cString: sqlite3_errmsg(db))
                    print("[VariableMapper] ❌ Failed to set all invalid: \(errorMsg)")
                }
            }
            sqlite3_finalize(statement)

            // 2. sortOrder順にlimit件を有効化
            let updateLimitQuery = """
                UPDATE variables
                SET valid = 1
                WHERE id IN (
                    SELECT id FROM variables
                    WHERE type = 'custom'
                    ORDER BY sortOrder ASC
                    LIMIT ?
                )
            """
            statement = nil

            if sqlite3_prepare_v2(db, updateLimitQuery, -1, &statement, nil) == SQLITE_OK {
                sqlite3_bind_int(statement, 1, Int32(limit))

                if sqlite3_step(statement) == SQLITE_DONE {
                    print("[VariableMapper] ✅ Set \(limit) custom variables to valid")
                } else {
                    let errorMsg = String(cString: sqlite3_errmsg(db))
                    print("[VariableMapper] ❌ Failed to set limit valid: \(errorMsg)")
                }
            }
            sqlite3_finalize(statement)

            print("[VariableMapper] ✅ updateValidFlags completed successfully")

        } catch {
            print("[VariableMapper] ❌ updateValidFlags failed: \(error)")
        }
    }
}
